File: eventbus/eventbus.go
package eventbus

import (
	"fmt"
	"../domain"
)

var subs = map[string][]func(domain.Event){}

func Publish(e domain.Event) {
    
	handlers := subs[e.Target]

    fmt.Printf("Published %v\n handlers: %v\n", e, len(handlers))

	for _, h := range handlers {

	    fmt.Printf("hanlder %v\n", h)
		go h(e)
	}
}

func Subscribe(name string, callback func(domain.Event)) {
    subs[name] = append(subs[name], callback)
}

File: CommandAndQueries/CommandAndQueries/EventBus.swift
//
//  File.swift
//  CommandAndQueries
//
//  Created by Vitaly Baum on 11/10/14.
//  Copyright (c) 2014 Vitaly Baum. All rights reserved.
//

import Foundation

struct EventBus {
    
    static var subscribtions:[(name: String, value: AnyObject -> ())] = []
    
    static func publish<T : Event>(e: T) {

        let y:String = NSStringFromClass(T.self)
        
        let handlers = subscribtions.filter({ x in x.name == y})
        
        for h in handlers {
            
            dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0)) {
                h.value(e)
            }
            
        }
    }
    
    static func subscribe<T : Event>(fun: T -> ()) {
    
        let y:String = NSStringFromClass(T.self)
        
        let fun2:(AnyObject)->() = {arg in fun(arg as T)}
        
        subscribtions.append(name: y, value: fun2)
    }
}
